### DISCUSSION TITLE
Animate when node changes

### DISCUSSION THREAD

**Discussion by potatochep** [xyflow/xyflow#4023] [Q&A]
I am currently using React Flow to show the data flow.

I am using custom nodes to dynamically add or delete nodes, and at this time, the automatic layout library elkjs is changing the xy coordinates of existing nodes.

Because of this, users are complaining of inconvenience, so I would like to give a simple animation when the location of the node changes.

Is it possible to do this?

I tried to use the css method such as transition, but this method failed because the dom element seems to be newly drawn every time the state value of the node changes.

**Comment by bcakmakoglu** [comment]
Technically possible but the library isn't going to provide you any built-in methods to "animate" the position change.

Something like this should work:
```tsx
  function animate(nodeId: string, position: XYPosition, duration: number) {
    const startTime = Date.now();

    const node = instance.getNode(nodeId);

    if (!node) {
      return;
    }

    const startingPosition = node.position;

    function frame() {
      const elapsed = Date.now() - startTime;
      const progress = Math.min(elapsed / duration, 1);

      const newX = interpolate(startingPosition.x, position.x, progress);
      const newY = interpolate(startingPosition.y, position.y, progress);

      instance.setNodes((nodes) =>
        nodes.map((node) => {
          if (node.id === nodeId) {
            return {
              ...node,
              position: {
                x: newX,
                y: newY,
              },
            };
          }

          return node;
        })
      );

      if (progress < 1) {
        requestAnimationFrame(frame);
      }
    }

    function interpolate(start: number, end: number, progress: number) {
      return start + (end - start) * progress;
    }

    requestAnimationFrame(frame);
  }
```

[Here's a sandbox demonstrating it.](https://codesandbox.io/p/sandbox/wispy-cache-mq57qt?file=%2Fsrc%2FApp.tsx%3A32%2C1-76%2C4)

**Comment by aetherwu** [comment]
However the offical pro demo animation looks seamless.
